### PR TITLE
Support multi-hash in multi-trie via PlainDB

### PR DESCRIPTION
**Pull request by sorpaas**
Requires https://github.com/paritytech/trie/pull/2, solves https://github.com/paritytech/substrate/issues/872#issuecomment-436816629

This adds multiple hash algorithm support for multi-trie. See https://github.com/paritytech/trie/pull/2's description for how this will work.

* `Ephemeral` implements `PlainDB` traits. `PlainDB` is `HashDB` with hashing stripped. Unlike `HashDB`, `PlainDB` will not check for null node, and it's expected that wrapper of `PlainDB` handles that.
* For `substrate-trie`, `hash_db::HashDB/PlainDB` are taken into function as generic (instead of trait objects). This change is because Rust does not support super-trait upcasting, but for functions like `child_delta_trie_root`, we need `DB` to be `HashDB + PlainDB`.

### CODE REVIEW

File: core/trie/src/lib.rs [hunk 1]
```diff
@@ -152,16 +166,22 @@ pub fn child_trie_root<H: Hasher, I, A, B>(_storage_key: &[u8], input: I) -> Vec
...
+pub fn child_delta_trie_root<H: Hasher, I, A, B, DB>(
+	_storage_key: &[u8],
+	db: &mut DB,
+	root_vec: Vec<u8>,
+	delta: I
+) -> Result<Vec<u8>, Box<TrieError<H::Out>>> where
 	I: IntoIterator<Item = (A, Option<B>)>,
 	A: AsRef<[u8]> + Ord,
 	B: AsRef<[u8]>,
+	DB: hash_db::HashDB<H, trie_db::DBValue> + hash_db::PlainDB<H::Out, trie_db::DBValue>,
```
Comment (sorpaas) [inline core/trie/src/lib.rs]: This looks like can't use the `HashDB` type alias defined above because Rust treats type alias and traits differently.

Comment (rphmeier) [reply core/trie/src/lib.rs]: there has been an RFC for trait aliases for ages but not happening anytime soon sadly. type aliases of traits are actually for the unsized values with the type of the trait.

**Comment by sorpaas** [comment]
Add two new traits in `hash-db`: `PlainDBRef` and `HashDBRef`. The reason for this is that sometimes we pass down immutable reference of `PlainDB`, and the user may want to wrap them and create new instance of `HashDB`. In this case, we need to implement immutable and mutable reference wrapper separately (because there's no generic over reference in Rust). For immutable reference, all the mutable functions still appear in the trait, and if without `PlainDBRef` and `HashDBRef`, they would need to be marked as `unreachable!()`, which is not nice.

**Comment by gavofyork** [comment]
would be great to get some more eyes on this as it may become relevant for a smart-contract chain with storage-rent.

@svyatonik will this present a big issue for light client and/or change trie?
@arkpar any interactions/significant maintenance burden for client/db?
@rphmeier any other issues?

**Comment by svyatonik** [comment]
For light client, imo, that wouldn't be a problem **iff** there'll be some strict correspondence (known to light client) between child storage (identified by `storage_key`) and the hashing algorithm that it uses. Would be good to see example of how it'll work for full client first, though. Right now I could see that child storage is [completely unavailable on light clients](https://github.com/paritytech/substrate/blob/d6808816291c8a20ed574f5c66b3f042c8069054/core/client/src/light/backend.rs#L331). This could be fixed easily (without multi hashing).

Changes trie is a separate trie, which could use its own hashing algorithm (i.e. Blake2 for all child storages). But there's another issue - we need to emit separate `DigestItem::ChangesTrieRoot` for every child storage in every block if we want they to be covered by changes tries. I'm not sure it is a good idea. Right now changes tries are only [covering top-level storage changes](https://github.com/paritytech/substrate/blob/d6808816291c8a20ed574f5c66b3f042c8069054/core/state-machine/src/changes_trie/build.rs#L78) - probably we should leave it as is?

**Comment by gavofyork** [comment]
> probably we should leave it as is?

Sounds reasonable, yeah.

@rphmeier @arkpar thoughts?

**Pull request merged**